tests/utils/state-locks-filter

```diff
@@ -0,0 +1,398 @@
+        curr_test.append(self.lines[start_line])
+
+        current_line = start_line + 1
+        while True:
+            if current_line >= len(self.lines) or self._is_test(
+                self.lines[current_line]
+            ):
+                return curr_test
+
+            curr_test.append(self.lines[current_line])
+            current_line = current_line + 1
+
+    # Retrieve the test lines
+    def get_test(self, test: str) -> list[str]:
+        for group in self.groups:
+            if test in group.get_name():
+                return str(group)
+
+        return ""
+
+    # Retrieve the times for each trace in the file
+    # For each trace, there is a list with the times for each test where the trace appears
+    def get_traces_times(self) -> dict[LockTrace, list[GroupTimes]]:
+        traces = dict[LockTrace, list[GroupTimes]]()
+        for group in self.groups:
+            group_traces = group.get_traces()
+            for trace in group_traces:
+                group_time = GroupTimes(
+                    group.get_name(),
+                    group.get_lock_held_time(trace),
+                    group.get_lock_wait_time(trace),
+                )
+                if not trace in traces.keys():
+                    traces[trace] = list[GroupTimes]()
+                traces.get(trace).append(group_time)
+
+        return traces
+
+
+def _make_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(description="spread filter helper")
+    parser.add_argument(
+        "-f", "--locks-file", type=str, required=True, help="Locks file"
+    )
+    parser.add_argument(
+        "--test",
+        default="",
+        help="Show results for this test, when test is selected no other filters are applied.",
+    )
+    parser.add_argument(
+        "--match",
+        action="append",
+        default=[],
+        help="Show traces that match this specific word",
+    )
+    parser.add_argument(
+        "--held-time", type=int, default=0, help="Include locks with longer held time"
+    )
+    parser.add_argument(
+        "--wait-time", type=int, default=0, help="Include locks with longer wait time"
+    )
+    parser.add_argument(
+        "--sort-held-time",
+        action="store_true",
+        help="Sort higher times first by held time",
+    )
+    parser.add_argument(
+        "--sort-wait-time",
+        action="store_true",
+        help="Sort higher times first by wait time",
+    )
+
+    return parser
+
+
+if __name__ == "__main__":
+    parser = _make_parser()
+    args = parser.parse_args()
+
+    if not os.path.isfile(args.locks_file):
+        print(
+            "state-lock-filter: locks file '{}' does not exist".format(args.locks_file)
+        )
+        sys.exit(1)
+
+    if args.sort_held_time and args.sort_wait_time:
+        print("state-lock-filter: define just 1 sorting (by held or wait times)")
+        sys.exit(1)
+
+    locks_reader = LocksFileReader(args.locks_file)
```

Comment:
Will this run after a test suite completes, or will we also be using this offline?

For each query, the trace file needs to be parsed again, and the query operators are only the ones we add ourselves. This leaves me wondering, what if as part of the test execution, we parse the file and dump all the data into sqlite artifact which we collect from the test? This would split the tool's functionality into 2 parts: parse & load into db, and run queries. We can have some basic queries implemented in the tool directly, but assuming some reasonable schema of db, there's nothing stopping us from processing this further or in another system.

Reply:
e.g. without any normalization:
```yaml
database:
  name: statelocks
  tables:
    - name: meta
      columns:
       # some sort of schema version
        - name: version
          type: integer
    - name: entries
      columns:
        - name: id
          type: integer
          primary_key: true
        - name: project
          type: string
        - name: group
          type: string
        - name: test
          type: string
          unique: true
        - name: held
          type: integer
        - name: wait
          type: integer
        - name: stracktrace
          type: text
```
populate, compress and attach as artifact

Reply:
I think this will be used offline, but then we could post process the results and send it to a db. @pedronis what do you prefer to do initially?

Reply:
let's try to land a first version without db, execute some tests runs and print the result of the grouping mode and then decide how to use this regularly